pkg/networkservice/common/connect/server.go

```diff
@@ -22,29 +22,34 @@ package connect
 import (
 	"context"
 	"net/url"
-
-	"github.com/networkservicemesh/sdk/pkg/tools/logger"
+	"sync"
 
 	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/pkg/errors"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
 
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clienturl"
-	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/inject/injecterror"
 	"github.com/networkservicemesh/sdk/pkg/tools/clientmap"
 	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
+	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+	"github.com/networkservicemesh/sdk/pkg/tools/logger"
 )
 
+type closeFunc func() bool
+
 type connectServer struct {
 	ctx               context.Context
 	clientFactory     func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient
 	clientDialOptions []grpc.DialOption
 	connInfos         connectionInfoMap
 	clients           clientmap.RefcountMap
+	clientsMutex      sync.Mutex
```

Comment:
NOTE: sync map will work here faster because in some cases we'll not block a group of clients.
Also, consider do you need order here. Because in this case you also could look at `serialize.Executor` for sync.

Reply:
Another connection with the same client can break it. So mutex is required here

Reply:
It looks like over synchronization.

Reply:
We have to be sure that other connection does not make write operation on the same client at the same time